File: examples/dekita/.claude/hooks/issue-review-response-check.py
#!/usr/bin/env python3
"""gh issue close時にAIレビューへの対応状況を確認してブロック。

Why:
    Issue作成後のAIレビューで改善提案が出ても、対応せずにクローズすると
    Issue品質が低下する。レビュー対応を強制することでIssue品質を維持する。

What:
    - gh issue closeコマンドを検出
    - IssueにAIレビューコメント（🤖 AI Review）があるか確認
    - レビュー後にIssue本文が更新されていなければブロック
    - スキップ環境変数（SKIP_REVIEW_RESPONSE）で回避可能

Remarks:
    - ブロック型フック
    - issue-ai-review.pyはレビュー実行、本フックはレビュー対応確認
    - コメントでの対応理由説明も有効な対応として扱う

Changelog:
    - silenvx/dekita#xxx: フック追加
    - silenvx/dekita#1024: SKIP_REVIEW_RESPONSE環境変数サポート
"""

import json
import os
import re
import subprocess
from datetime import datetime

from lib.constants import TIMEOUT_MEDIUM
from lib.execution import log_hook_execution
from lib.results import make_block_result, print_continue_and_log_skip
from lib.session import create_hook_context, parse_hook_input
from lib.strings import extract_inline_skip_env, is_skip_env_enabled, strip_quoted_strings

SKIP_REVIEW_RESPONSE_ENV = "SKIP_REVIEW_RESPONSE"


def extract_issue_number(command: str) -> str | None:
    """Extract issue number from gh issue close command."""
    # Remove quoted strings to avoid false positives
    cmd = strip_quoted_strings(command)

    # Check if this is a gh issue close command
    if not re.search(r"gh\s+issue\s+close\b", cmd):
        return None

    # Extract all arguments after "gh issue close"
    match = re.search(r"gh\s+issue\s+close\s+(.+)", cmd)
    if not match:
        return None

    args = match.group(1)

    # Find issue number (with or without #) among the arguments
    for part in args.split():
        if part.startswith("-"):
            continue
        num_match = re.match(r"#?(\d+)$", part)
        if num_match:
            return num_match.group(1)

    return None


def get_ai_review_comment_time(issue_number: str) -> datetime | None:
    """Get the timestamp of AI Review comment if exists.

    Returns:
        datetime of AI Review comment, or None if not found.
    """
    try:
        result = subprocess.run(
            [
                "gh",
                "issue",
                "view",
                issue_number,
                "--json",
                "comments",
                "--jq",
                '.comments[] | select(.body | contains("🤖 AI Review")) | .createdAt',
            ],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
            check=False,
        )

        if result.returncode != 0 or not result.stdout.strip():
            return None

        # Get the latest (newest) AI Review comment to ensure
        # edits happened after the most recent review
        timestamps = result.stdout.strip().split("\n")
        if timestamps:
            # Parse ISO format: 2025-12-20T12:05:59Z
            # Use the last timestamp (newest review)
            return datetime.fromisoformat(timestamps[-1].replace("Z", "+00:00"))

        return None

    except (subprocess.TimeoutExpired, OSError, ValueError):
        return None


def was_issue_edited_after(issue_number: str, after_time: datetime) -> bool:
    """Check if issue was updated after the given time.

    Uses issue's updated_at field to detect any activity (body edits,
    comments, label changes, etc.) after the AI Review.

    Note: This intentionally treats comments as valid responses,
    since the hook allows "explaining why a suggestion is not needed"
    via comments as a valid way to address AI Review feedback.
    """
    try:
        # Get issue updated_at timestamp directly
        result = subprocess.run(
            [
                "gh",
                "api",
                f"repos/:owner/:repo/issues/{issue_number}",
                "--jq",
                ".updated_at",
            ],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
            check=False,
        )

        if result.returncode != 0:
            # API error, don't block
            return True

        if not result.stdout.strip():
            # No updated_at found, don't block
            return True

        try:
            updated_at = datetime.fromisoformat(result.stdout.strip().replace("Z", "+00:00"))
            # Issue was updated after AI Review comment
            return updated_at > after_time
        except ValueError:
            # Parse error, don't block
            return True

    except (subprocess.TimeoutExpired, OSError):
        # On error, don't block
        return True


def get_ai_review_suggestions(issue_number: str) -> list[str]:
    """Extract bullet point suggestions from AI Review comment.

    Parses the AI Review comment body and extracts lines starting with
    '- ' or '* ' as improvement suggestions. Returns up to 3 suggestions,
    each truncated to 100 characters.

    Note: This extracts bullet points as a heuristic for finding actionable
    suggestions. Not all bullet points may be actual suggestions, but this
    provides helpful context for the user.
    """
    try:
        result = subprocess.run(
            [
                "gh",
                "issue",
                "view",
                issue_number,
                "--json",
                "comments",
                "--jq",
                '.comments[] | select(.body | contains("🤖 AI Review")) | .body',
            ],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
            check=False,
        )

        if result.returncode != 0 or not result.stdout.strip():
            return []

        body = result.stdout.strip()

        # Extract improvement suggestions (lines starting with - or *)
        suggestions = []
        for line in body.split("\n"):
            line = line.strip()
            if line.startswith("- ") or line.startswith("* "):
                # Skip very short suggestions (< 10 chars are likely not meaningful)
                if len(line) > 10:
                    suggestions.append(line[:100])  # Truncate long lines
                    if len(suggestions) >= 3:
                        break

        return suggestions

    except (subprocess.TimeoutExpired, OSError):
        return []


def main():
    """PreToolUse hook for Bash commands."""
    result = {"decision": "approve"}

    try:
        input_data = parse_hook_input()
        # Issue #2607: Create context for session_id logging
        ctx = create_hook_context(input_data)
        tool_name = input_data.get("tool_name", "")

        if tool_name != "Bash":
            print_continue_and_log_skip(
                "issue-review-response-check", f"not Bash: {tool_name}", ctx=ctx
            )
            return

        tool_input = input_data.get("tool_input", {})
        command = tool_input.get("command", "")

        # Check if this is a gh issue close command
        issue_number = extract_issue_number(command)

        # Issue #1024: SKIP_REVIEW_RESPONSE environment variable support
        # Allows bypassing AI review response check for confirmed cases
        # Supports both exported env var and inline (SKIP_REVIEW_RESPONSE=1 gh issue close)
        if issue_number:
            # Check exported environment variable
            if is_skip_env_enabled(os.environ.get(SKIP_REVIEW_RESPONSE_ENV)):
                log_hook_execution(
                    "issue-review-response-check",
                    "skip",
                    f"SKIP_REVIEW_RESPONSE=1: Issue #{issue_number} のAIレビュー対応チェックをスキップ",
                )
                print(json.dumps(result))
                return
            # Check inline environment variable
            inline_value = extract_inline_skip_env(command, SKIP_REVIEW_RESPONSE_ENV)
            if is_skip_env_enabled(inline_value):
                log_hook_execution(
                    "issue-review-response-check",
                    "skip",
                    f"SKIP_REVIEW_RESPONSE=1: Issue #{issue_number} のAIレビュー対応チェックをスキップ（インライン）",
                )
                print(json.dumps(result))
                return
        if not issue_number:
            print_continue_and_log_skip(
                "issue-review-response-check", "no issue number found", ctx=ctx
            )
            return

        # Check for AI Review comment
        ai_review_time = get_ai_review_comment_time(issue_number)

        if not ai_review_time:
            # No AI Review, let it through
            log_hook_execution(
                "issue-review-response-check",
                "approve",
                f"AIレビューなし: Issue #{issue_number}",
            )
            print(json.dumps(result))
            return

        # Check if issue was edited after AI Review
        if was_issue_edited_after(issue_number, ai_review_time):
            log_hook_execution(
                "issue-review-response-check",
                "approve",
                f"AIレビュー後に編集あり: Issue #{issue_number}",
            )
            print(json.dumps(result))
            return

        # Issue has AI Review but was not edited - block
        suggestions = get_ai_review_suggestions(issue_number)
        suggestion_text = ""
        if suggestions:
            suggestion_text = "\n\n**AIレビューの改善提案例:**\n" + "\n".join(suggestions)

        reason = (
            f"⚠️ Issue #{issue_number} にはAIレビューコメントがありますが、"
            f"レビュー後にIssue本文が更新されていません。\n\n"
            f"**対応方法:**\n"
            f"1. `gh issue view {issue_number} --comments` でAIレビューを確認\n"
            f"2. 改善提案をIssue本文に反映（`gh issue edit {issue_number}`）\n"
            f"3. 対応不要な提案は、その理由をコメントに記載\n"
            f"4. その後、再度クローズを実行"
            f"{suggestion_text}"
        )
        result = make_block_result("issue-review-response-check", reason)

        log_hook_execution(
            "issue-review-response-check",
            "block",
            f"AIレビュー未対応: Issue #{issue_number}",
        )

    except Exception as e:
        # Don't block on errors - reset to approve
        result = {"decision": "approve"}
        log_hook_execution(
            "issue-review-response-check",
            "error",
            f"フックエラー: {e}",
        )

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/issue_review_response_check.ts
#!/usr/bin/env bun
/**
 * gh issue close時にAIレビューへの対応状況を確認してブロック。
 *
 * Why:
 *   Issue作成後のAIレビューで改善提案が出ても、対応せずにクローズすると
 *   Issue品質が低下する。レビュー対応を強制することでIssue品質を維持する。
 *
 * What:
 *   - gh issue closeコマンドを検出
 *   - IssueにAIレビューコメント（🤖 AI Review）があるか確認
 *   - レビュー後にIssue本文が更新されていなければブロック
 *   - スキップ環境変数（SKIP_REVIEW_RESPONSE）で回避可能
 *
 * Remarks:
 *   - ブロック型フック
 *   - 過去に実行されたAIレビュー（🤖 AI Review）への対応確認
 *   - コメントでの対応理由説明も有効な対応として扱う
 *   - Python版: issue_review_response_check.py
 *
 * Changelog:
 *   - silenvx/dekita#xxx: フック追加
 *   - silenvx/dekita#1024: SKIP_REVIEW_RESPONSE環境変数サポート
 *   - silenvx/dekita#2917: TypeScript版初期実装
 */

import { execSync } from "node:child_process";
import { TIMEOUT_MEDIUM } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { makeApproveResult, makeBlockResult } from "../lib/results";
import { parseHookInput } from "../lib/session";
import { extractInlineSkipEnv, isSkipEnvEnabled, stripQuotedStrings } from "../lib/strings";

const HOOK_NAME = "issue-review-response-check";
const SKIP_ENV_NAME = "SKIP_REVIEW_RESPONSE";

/**
 * Extract issue number from gh issue close command.
 */
export function extractIssueNumber(command: string): string | null {
  const cmd = stripQuotedStrings(command);

  if (!/gh\s+issue\s+close\b/.test(cmd)) {
    return null;
  }

  const match = cmd.match(/gh\s+issue\s+close\s+(.+)/);
  if (!match) {
    return null;
  }

  const args = match[1];

  for (const part of args.split(/\s+/)) {
    if (part.startsWith("-")) {
      continue;
    }
    const numMatch = part.match(/^#?(\d+)$/);
    if (numMatch) {
      return numMatch[1];
    }
  }

  return null;
}

/**
 * Get the timestamp of AI Review comment if exists.
 */
function getAiReviewCommentTime(issueNumber: string): Date | null {
  try {
    const result = execSync(
      `gh issue view ${issueNumber} --json comments --jq '.comments[] | select(.body | contains("🤖 AI Review")) | .createdAt'`,
      {
        encoding: "utf-8",
        timeout: TIMEOUT_MEDIUM * 1000,
        stdio: ["pipe", "pipe", "pipe"],
      },
    );

    if (!result.trim()) {
      return null;
    }

    // Get the latest (newest) AI Review comment
    const timestamps = result.trim().split("\n");
    if (timestamps.length > 0) {
      const lastTimestamp = timestamps[timestamps.length - 1];
      return new Date(lastTimestamp);
    }

    return null;
  } catch {
    return null;
  }
}

/**
 * Check if issue was updated after the given time.
 *
 * Uses issue's updated_at field to detect any activity after the AI Review.
 * This intentionally treats comments as valid responses.
 */
function wasIssueEditedAfter(issueNumber: string, afterTime: Date): boolean {
  try {
    const result = execSync(`gh api repos/:owner/:repo/issues/${issueNumber} --jq '.updated_at'`, {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });

    if (!result.trim()) {
      // No updated_at found, don't block
      return true;
    }

    const updatedAt = new Date(result.trim());
    return updatedAt > afterTime;
  } catch {
    // On error, don't block
    return true;
  }
}

/**
 * Extract bullet point suggestions from AI Review comment.
 */
function getAiReviewSuggestions(issueNumber: string): string[] {
  try {
    const result = execSync(
      `gh issue view ${issueNumber} --json comments --jq '.comments[] | select(.body | contains("🤖 AI Review")) | .body'`,
      {
        encoding: "utf-8",
        timeout: TIMEOUT_MEDIUM * 1000,
        stdio: ["pipe", "pipe", "pipe"],
      },
    );

    if (!result.trim()) {
      return [];
    }

    const body = result.trim();
    const suggestions: string[] = [];

    for (const line of body.split("\n")) {
      const trimmed = line.trim();
      if (trimmed.startsWith("- ") || trimmed.startsWith("* ")) {
        // Skip very short suggestions
        if (trimmed.length > 10) {
          suggestions.push(trimmed.slice(0, 100));
          if (suggestions.length >= 3) {
            break;
          }
        }
      }
    }

    return suggestions;
  } catch {
    return [];
  }
}

async function main(): Promise<void> {
  let sessionId: string | undefined;
  try {
    const data = await parseHookInput();
    sessionId = data.session_id;
    const toolName = (data.tool_name as string) || "";

    // Only check Bash commands
    if (toolName !== "Bash") {
      await logHookExecution(HOOK_NAME, "approve", `not Bash: ${toolName}`, undefined, {
        sessionId,
      });
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    const toolInput = (data.tool_input as Record<string, unknown>) || {};
    const command = (toolInput.command as string) || "";

    // Check if this is a gh issue close command
    const issueNumber = extractIssueNumber(command);

    // Check for skip environment variable
    if (issueNumber) {
      if (isSkipEnvEnabled(process.env[SKIP_ENV_NAME])) {
        await logHookExecution(
          HOOK_NAME,
          "approve",
          `SKIP_REVIEW_RESPONSE=1: Issue #${issueNumber} のAIレビュー対応チェックをスキップ`,
          undefined,
          { sessionId },
        );
        const result = makeApproveResult(HOOK_NAME);
        console.log(JSON.stringify(result));
        return;
      }

      const inlineValue = extractInlineSkipEnv(command, SKIP_ENV_NAME);
      if (isSkipEnvEnabled(inlineValue)) {
        await logHookExecution(
          HOOK_NAME,
          "approve",
          `SKIP_REVIEW_RESPONSE=1: Issue #${issueNumber} のAIレビュー対応チェックをスキップ（インライン）`,
          undefined,
          { sessionId },
        );
        const result = makeApproveResult(HOOK_NAME);
        console.log(JSON.stringify(result));
        return;
      }
    }

    if (!issueNumber) {
      await logHookExecution(HOOK_NAME, "approve", "no issue number found", undefined, {
        sessionId,
      });
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    // Check for AI Review comment
    const aiReviewTime = getAiReviewCommentTime(issueNumber);

    if (!aiReviewTime) {
      // No AI Review, let it through
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `AIレビューなし: Issue #${issueNumber}`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    // Check if issue was edited after AI Review
    if (wasIssueEditedAfter(issueNumber, aiReviewTime)) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `AIレビュー後に編集あり: Issue #${issueNumber}`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    // Issue has AI Review but was not edited - block
    const suggestions = getAiReviewSuggestions(issueNumber);
    let suggestionText = "";
    if (suggestions.length > 0) {
      suggestionText = `\n\n**AIレビューの改善提案例:**\n${suggestions.join("\n")}`;
    }

    const reasonLines = [
      `⚠️ Issue #${issueNumber} にはAIレビューコメントがありますが、`,
      "レビュー後にIssue本文が更新されていません。",
      "",
      "**対応方法:**",
      `1. \`gh issue view ${issueNumber} --comments\` でAIレビューを確認`,
      `2. 改善提案をIssue本文に反映（\`gh issue edit ${issueNumber}\`）`,
      "3. 対応不要な提案は、その理由をコメントに記載",
      "4. その後、再度クローズを実行",
    ];

    const blockMessage = reasonLines.join("\n") + suggestionText;

    await logHookExecution(
      HOOK_NAME,
      "block",
      `AIレビュー未対応: Issue #${issueNumber}`,
      undefined,
      { sessionId },
    );

    const result = makeBlockResult(HOOK_NAME, blockMessage);
    console.log(JSON.stringify(result));
    process.exit(2);
  } catch (error) {
    // Don't block on errors - approve silently
    await logHookExecution(HOOK_NAME, "error", `フックエラー: ${formatError(error)}`, undefined, {
      sessionId,
    });
    const result = makeApproveResult(HOOK_NAME);
    console.log(JSON.stringify(result));
  }
}

if (import.meta.main) {
  main();
}
